src/extract_gpuinfo_amdgpu.c

```diff
@@ -352,6 +365,46 @@ static void authenticate_drm(int fd) {
   fprintf(stderr, "Failed to authenticate to DRM; XCB authentication unimplemented\n");
 }
 
+static void initDeviceSysfsPaths(struct gpu_info_amdgpu *gpu_info) {
+  // Open the device sys folder to gather information not available through the DRM driver
+  char devicePath[22 + PDEV_LEN];
+  snprintf(devicePath, sizeof(devicePath), "/sys/bus/pci/devices/%s", gpu_info->pdev);
+  gpu_info->sysfsFD = open(devicePath, O_RDONLY);
+
+  // Open the device hwmon folder (Fan speed are available there)
+  static const char hwmon[] = "hwmon";
+  if (gpu_info->sysfsFD >= 0) {
+    int hwmondirFD = openat(gpu_info->sysfsFD, hwmon, O_RDONLY);
+    if (hwmondirFD >= 0) {
+      DIR *hwmonDir = fdopendir(hwmondirFD);
+      if (hwmonDir) {
+        struct dirent *dirEntry;
+        while ((dirEntry = readdir(hwmonDir))) {
+          // There should be one directory inside hwmon, with a name having the following pattern hwmon[0-9]+
+          if (dirEntry->d_type == DT_DIR) {
+            size_t matchLen = 0;
+            for (matchLen = 0; matchLen < sizeof(hwmon) - 1; ++matchLen) {
+              if (dirEntry->d_name[matchLen] == '\0' || hwmon[matchLen] != dirEntry->d_name[matchLen])
+                break;
+            }
+            // We found our candidate
+            if (matchLen == sizeof(hwmon) - 1)
+              break;
+          }
+        }
+        if (dirEntry) {
+          gpu_info->hwmonFD = openat(dirfd(hwmonDir), dirEntry->d_name, O_RDONLY);
+        } else
+          gpu_info->hwmonFD = -1;
```

Comment:
How about instead of adding these else-s, pre-initialize `gpu_info->hwmonFD = -1;` and only assign it if things go well? Could also avoid nested ifs by returning early.

Reply:
You are right.
I do the convoluted if/else spaghetti as a first pass when the code is still changing many times.
I will clean this up.